### ISSUE TITLE
Hope to get some clarification on the reward model

### ISSUE THREAD

**Issue by disperaller** [ashishjamarkattel/reinforment-learning-with-human-feedback#5]
in the notebook rhlf.ipynb, i hope someone could help me clarify some points. 

When defining the reward model, the code use the following:
```
tokenizer = AutoTokenizer.from_pretrained(MODEL_PATH)
model = AutoModelForCausalLM.from_pretrained(MODEL_PATH)
```

However, as far as i know, maybe i was wrong, but shouldn't the model only output 1 scalar value serves as the reward score for a particular response? In order to do that, the model should be loaded by using something like AutoModel....ForSequenceClassification, or AutoModel...WithValueHead, correct? Here the AutoModelForCausalLM outputs of size (batch_size, seq_len, vocab), which in my understanding, is incorrect to use.

I was suspicious with this setting, but still carry on to the following code, until i get to the get_score(model, tokenizer, prompt, response) function. Theoretically this function should serve as a method to compute the score based on the prediction. However, the function outputs logits, which is of size (batch_size, seq_len, vocab) as mentioned above. This logits, lets say logitsWin for chosen response, is assigned to loss1, logitsLose for rejected response, is assigned to loss2:
```
loss1 = get_score(model, tokenizer, prompt, example_prefered_response)
loss2= get_score(model, tokenizer, prompt, example_unprefered_response)
```

Later, according to this line of code:
`loss = -nn.functional.logsigmoid(loss1 - loss2).mean()`
the two 'losses' get subtracted on each element, which still result in the shape of (b, s, v). Then the loss is computed directly by taking the logsigmoid ON EVERY ELEMENT in the subtraction result, and then average all to get the final loss. If this is correct, then the loss is computed using every predicted token's probability over vocabulary? To my knowledge, this is incorrect. The reward should only be computed by leveraging the distance between winning reward score and losing reward score, which should be generated by looking at the entire sentence. In SequenceClassification's case, it should only be the [CLS] token, not the remaining tokens.

However, since the code runs with no error, I suspect I was wrong,  could anyone help me on this?

**Comment by ashishjamarkattel** [comment]
Yes you are right. Generally it is prefered to use automodel for sequenceclassifcation with label to be scalar vector. 
but taking the vector we are doing same. Which is generally giving the score for 2 vector distance. just like we do while calculating  sementic search. so getting mean is only score that define the model score on how the much is choosen away from rejected. 

As per my knowledge it doesnot matter, any how we are making loss of chosen far from rejected.

**Comment by disperaller** [comment]
> Yes you are right. Generally it is prefered to use automodel for sequenceclassifcation with label to be scalar vector. but taking the vector we are doing same. Which is generally giving the score for 2 vector distance. just like we do while calculating sementic search. so getting mean is only score that define the model score on how the much is choosen away from rejected.
> 
> As per my knowledge it doesnot matter, any how we are making loss of chosen far from rejected.

ok thank you for the clarification :)
